### PR TITLE
Add base for CUDA allReduce and broadcast in DataChannelGloo

### PR DESCRIPTION
**Pull request by apaszke**

### CODE REVIEW

File: torch/lib/THD/base/data_channels/GlooCache.hpp [hunk 1]
```diff
@@ -126,14 +146,46 @@ struct GlooCache {
...
+
+  static void memcpy_output(value_type& info, thpp::Tensor& t) {
+    std::uint64_t tensor_bytes = t.elementSize() * t.numel();
+    auto t_dev = getDeviceType(t);
+    auto output_buffer = GlooCache::output_buffer(info).get();
+
+    if (t_dev == DeviceType::CPU) {
+      std::memcpy(t.data(), output_buffer, tensor_bytes);
+    } else if (t_dev == DeviceType::CUDA) {
+      THCudaCheck(cudaMemcpy(t.data(), output_buffer, tensor_bytes, cudaMemcpyDeviceToDevice));
```
Comment (VirrageS) [inline torch/lib/THD/base/data_channels/GlooCache.hpp]: Not protected by `#ifdef WITH_CUDA`

File: torch/lib/THD/base/data_channels/GlooCache.hpp [hunk 2]
```diff
@@ -126,14 +146,46 @@ struct GlooCache {
...
 
+  static void memcpy_input(value_type& info, thpp::Tensor& t) {
+    std::uint64_t tensor_bytes = t.elementSize() * t.numel();
+    auto t_dev = getDeviceType(t);
+    auto input_buffer = GlooCache::input_buffer(info).get();
+
+    if (t_dev == DeviceType::CPU) {
+      std::memcpy(input_buffer, t.data(), tensor_bytes);
+    } else if (t_dev == DeviceType::CUDA) {
+      THCudaCheck(cudaMemcpy(input_buffer, t.data(), tensor_bytes, cudaMemcpyDeviceToDevice));
```
Comment (VirrageS) [inline torch/lib/THD/base/data_channels/GlooCache.hpp]: Not protected by `#ifdef WITH_CUDA`

File: torch/lib/THD/base/data_channels/GlooCache.hpp [hunk 3]
```diff
@@ -8,22 +8,34 @@
...
+#include <cuda.h>
+#include <THC/THC.h>
+#endif
 #include <cstdint>
 #include <memory>
 #include <tuple>
 #include <vector>
 
+// TODO: it should be possible to set this using the library API...
+extern THCState *state;
```
Comment (VirrageS) [inline torch/lib/THD/base/data_channels/GlooCache.hpp]: Not protected by `#ifdef WITH_CUDA`

File: torch/lib/THD/base/data_channels/GlooCache.hpp [hunk 4]
```diff
@@ -100,9 +112,17 @@ struct GlooCache {
...
-  std::shared_ptr<buffer_type> createBuffer(std::size_t bytes) const {
-    return std::shared_ptr<buffer_type>(new char[bytes],
-                                        std::default_delete<char[]>());
+  std::shared_ptr<buffer_type> createBuffer(std::size_t bytes, DeviceType device) const {
+    if (device == DeviceType::CPU) {
+      return std::shared_ptr<buffer_type>(new char[bytes],
+                                          std::default_delete<char[]>());
+    } else if (device == DeviceType::CUDA) {
+      buffer_type *buf;
+      THCudaCheck(THCudaMalloc(state, (void**)&buf, bytes));
```
Comment (VirrageS) [inline torch/lib/THD/base/data_channels/GlooCache.hpp]: Not protected by `#ifdef WITH_CUDA`

**Comment by apaszke** [comment]
Good point, I've added the guards!

**Review by VirrageS — APPROVED** [review]

**Pull request merged**